### PR TITLE
Reprocessed Modal: Submission History

### LINKED ISSUE #3271: Reprocessed on enhancement (3014 follow-on)
**Description:**
Adds improvement to Submitted On column of submission history replacing current most recent reprocessed date with a clickable trigger that opens a dialog containing a full list of reprocessing event timestamps. 

**Acceptance Criteria:** 

 
- [x] Submitted on/by columns are combined
- [ ] Dialog has a close button
- [ ] Dialog contains paragraph explainer on reprocessing
- [ ] Dialog contains date time stamp for last reprocessing event
- [ ] Dialog opens when trigger in Submitted on column is clicked
- [ ] UX has been pinged to ticket knowledge center revisions to capture these changes
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**

- [x] Add dialog
- [x] Add dialog trigger
- [x] Add explainer text
- [x] Add timestamp
- [x] Combine Submitted On/By columns
- [ ] A11y test in raft review
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
Reprocessed on trigger example pending mockups:

[Reprocessed ⓘ]()

Dialog contents:

---
We've reprocessed your files with updated validation criteria, based on system improvements, to improve accuracy of error reports. No changes have been made to your original data submission. 

Data was reprocessed on: [Datetime]

[Close]()

---

**Supporting Documentation:**

image

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Add ReprocessedModal component
- Use new component in submission history tables
Pull request closes #3271 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Submit some files.
3. Note that the `Reprocessed` button isn't visible in submission history
4. Reparse some/all of the files you submitted
5. You should now see the `Reprocessed` button is visible in submission history for the files you reparsed
6. Click the button and you should see a modal with the latest reparse date and explanation

In the `raft` environment use the following STT, year, and quarter to easily view the changes: `Arizona Year:2021 and Quarter:1`

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Submitted on/by columns are combined
- [x] Dialog has a close button
- [x] Dialog contains paragraph explainer on reprocessing
- [x] Dialog contains date time stamp for last reprocessing event
- [x] Dialog opens when trigger in Submitted on column is clicked
- [x] UX has been pinged to ticket knowledge center revisions to capture these changes
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [x] Does PR include documentation for Raft's a11y review? 
+ [x] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Review by victoriaatraft — APPROVED** [review]
Screenshot 2024-12-10 at 10 44 14 AM
LGTM from an a11y perspective. @elipe17. Is it possible for the modal to also close of the user selects the grayed out area in addition to the "close" button?

**Review by andrew-jameson — APPROVED** [review]

File: tdrs-frontend/src/components/SubmissionHistory/CaseAggregatesTable.jsx
Comment (jtimpe) [inline tdrs-frontend/src/components/SubmissionHistory/CaseAggregatesTable.jsx]: So this doesn't break anything because of how the modal's css/javascript work under the hood, but it is technically against DOM structure best practices to have an element appear visually "on top" but be nested inside other elements on the page. What you have would be how I would implement a tooltip component, but I'd recommend moving the modal to be a direct child of the fragment in `Reports.jsx` and letting the parent container own all the state, passing the `setState` function as a callback.

```jsx
export const Reports = () => {
  const [reprocessedModalVisible, setReprocessedModalVisible] = useState(false)
  // ......

  return (
    <>
      <div>
        ......
        <CaseAggregatesTable setReprocessedModalVisible={setReprocessedModalVisible} />
      </div>
      <Modal />
      <ReprocessedModal isVisible={reprocessedModalVisible} />
    </>
  )
}
```

`setReprocessedModalVisible` needs to be passed around a few times and may need to take some information like the reprocessed date (or the submission id, and the date can be fetched by the component). But would keep the page from creating a separate modal for each row. 

This line would then be changed to only contain the button, which executes the `setReprocessedModalVisible` callback

File: tdrs-frontend/src/components/SubmissionHistory/ReprocessedModal.jsx
Comment (jtimpe) [inline tdrs-frontend/src/components/SubmissionHistory/ReprocessedModal.jsx]: related to my other comment but wanted to call this out specifically - it's generally advised to keep state as high up in the component chain as makes sense. this goes along with having a single modal controlled by the root page. 
- https://legacy.reactjs.org/docs/lifting-state-up.html

**Review by jtimpe — COMMENTED** [review]
Works as written! I have a suggestion regarding the structuring of the components that I would like your thoughts on before I approve, but I'll let you decide whether to move forward with it or not.

File: tdrs-frontend/src/components/SubmissionHistory/CaseAggregatesTable.jsx
Comment (elipe17) [reply tdrs-frontend/src/components/SubmissionHistory/CaseAggregatesTable.jsx]: Oh that is something I never considered. It did not occur to me that I was actually creating a modal per row in the tables. I will definitely make that change, thank you for the suggestion!

Comment (jtimpe) [reply tdrs-frontend/src/components/SubmissionHistory/CaseAggregatesTable.jsx]: @elipe17 @reitermb also while i'm thinking of it, tooltip might be available to us again after the v3 upgrade
https://designsystem.digital.gov/components/tooltip/

**Comment by elipe17** [comment]
@jtimpe let me know what you think about the changes

File: tdrs-frontend/src/components/Modal/Modal.jsx
Comment (elipe17) [inline tdrs-frontend/src/components/Modal/Modal.jsx:30]: This fixed a bug where we could get an array out of bounds error. We could also escape the modal with tabs when this error occurred.

**Review by jtimpe — APPROVED** [review]
Good work!

**Review by raftmsohani — APPROVED** [review]
LGTM!

**Review by reitermb — APPROVED** [review]

File: tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js
Comment (ADPennington) [inline tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js:490]: @elipe17 do we need more tests here to capture the following scenarios:

- reprocessing tribal TANF file
- reprocessing SSP file
- no reprocessing 

Asking because I noticed that reprocessing indicators are not appearing for tribal TANF and SSP files. evidence below. 

Screenshot 2024-12-16 124333
Screenshot 2024-12-16 123500

**Review by ADPennington — CHANGES_REQUESTED** [review]
changes requested re: reprocessing on tribal and SSP files.

File: tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js
Comment (elipe17) [reply tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js:490]: It should be the same exact logic regardless of TANF, SSP, Tribal. I just tested SSP locally and it was fine. Let me checkout the deployed env

Comment (elipe17) [reply tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js:490]: @ADPennington I just tested in the Raft env since it is still deployed there and it is working as expected for TANF, SSP, and Tribal. Maybe we can deploy this back to qasp when you are ready and check it out together?

Comment (ADPennington) [reply tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js:490]: thank you @elipe17 -- one thing i forgot to mention is that I was testing aggregate and stratum files for SSP and tribal TANF.

Comment (elipe17) [reply tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js:490]: Roger. I tested active, closed, and aggregate for TANF. Active and aggregate for SPP. Only active for Tribal. Let me know when you are ready to test it out and we can hop on a call or work async via the deployment

Comment (ADPennington) [reply tdrs-frontend/src/components/SubmissionHistory/SubmissionHistory.test.js:490]: @elipe17 re-deploying now

**Comment by ADPennington** [comment]
@ttran-hub -- do you have some time today for a11y testing for this PR? 

we need a11y testing on data file page >  submission history tab, in the areas highlighted in the screenshot below. This branch is in [qasp env](https://tdp-frontend-qasp.app.cloud.gov).

Screenshot 2024-12-17 114759

@klinkoberstar can you remind me -- is there a release note associated with this?

**Comment by ttran-hub** [comment]
> @ttran-hub -- do you have some time today for a11y testing for this PR?
> 
> we need a11y testing on data file page > submission history tab, in the areas highlighted in the screenshot below. This branch is in [qasp env](https://tdp-frontend-qasp.app.cloud.gov).
> 
> Screenshot 2024-12-17 114759
> 
> @klinkoberstar can you remind me -- is there a release note associated with this?

a111y LGTM!

**Review by ADPennington — APPROVED** [review]
🚀 🚀 🚀

**Pull request merged**
